src/pkg/library/client/push.go

```diff
@@ -40,284 +32,111 @@ func UploadImage(filePath string, libraryRef string, libraryURL string) error {
 	}
 	sylog.Debugf("Image hash computed as %s\n", imageHash)
 
-	entity, collection, container, tags := parseLibraryRef(libraryRef)
+	entityName, collectionName, containerName, tags := parseLibraryRef(libraryRef)
 
-	entityID, err := entityExists(entity)
+	// Find or create entity
+	entity, found, err := getEntity(libraryURL, entityName)
 	if err != nil {
 		return err
 	}
-	if entityID == "" {
-		sylog.Verbosef("Entity %s does not exist in library - creating it.\n", entity)
-		entityID, err = createEntity(entity)
+	if !found {
+		sylog.Verbosef("Entity %s does not exist in library - creating it.\n", entityName)
+		entity, err = createEntity(libraryURL, entityName)
 		if err != nil {
 			return err
 		}
 	}
-	collectionID, err := collectionExists(entity, collection)
-	if err != nil {
-		return err
-	}
-	if collectionID == "" {
-		sylog.Verbosef("Collection %s/%s does not exist in library - creating it.\n", entity, collection)
-		collectionID, err = createCollection(collection, entityID)
-		if err != nil {
-			return err
-		}
-	}
-	containerID, err := containerExists(entity, collection, container)
+
+	// Find or create collection
+	collection, found, err := getCollection(libraryURL, entityName+"/"+collectionName)
 	if err != nil {
 		return err
 	}
-	if containerID == "" {
-		sylog.Verbosef("Container %s/%s/%s does not exist in library - creating it.\n", entity, collection, container)
-		containerID, err = createContainer(container, collectionID)
+	if !found {
+		sylog.Verbosef("Collection %s does not exist in library - creating it.\n", collectionName)
```

Comment:
Are newlines at the end really required when using `sylog`? It seems counter to convention in all other Go loggers I've come across. Thoughts? Should we file an issue to get this looked at? I feel like we'll forever be chasing down missing newlines otherwise... and given the logger is line-based, I can't see a case where this behaviour would not be desired?

Reply:
Yes, newlines are required for sylog. I believe @bauerm97 wanted to replicate the behaviour of the current Singularity C logging